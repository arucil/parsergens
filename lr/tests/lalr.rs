use insta::assert_snapshot;

mod parse;

#[test]
fn precedence() {
  let parser = lr::build(r#"
%token plus "+"
%token minus "-"
%token mult "*"
%token div "/"
%token pow "^"
%token equal "=="
%token lparen "("
%token rparen ")"
%token comma ","
%token num /\d+/
%token id /[a-zA-Z]\w*/

%skip /[ \t\n]+/

%non-assoc equal
%left-assoc plus minus
%left-assoc mult div
%right-assoc pow
%right-assoc NEG

%start E

E = E plus E
  | E minus E
  | E mult E
  | E div E
  | E pow E
  | E equal E
  | minus E     %prec NEG
  | lparen E rparen
  | id lparen ARGS rparen
  | num
  | id

ARGS = ()
  | E (comma E)*
  "#, lr::ParserKind::Lalr).unwrap();

  let input = "12 + 3 * x / 5 - 50 - foo ^ -y ^ (-1 == f() + g(3*-4, (p)))";

  assert_snapshot!(parse::parse(&parser, input, "E").join("\n"));
}
