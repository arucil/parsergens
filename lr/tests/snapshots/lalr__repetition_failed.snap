---
source: lr/tests/lalr.rs
expression: "parse::parse(&parser, input2, \"S\").join(\"\\n\")"
---
shift  a
reduce S -> a
error token a at 1:2, expected EOF
